sos/report/plugins/mellanox_firmware.py

```diff
@@ -0,0 +1,126 @@
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+from sos.report.plugins import Plugin, SoSPredicate, IndependentPlugin
+import os
+import time
+
+
+class MellanoxFirmware(Plugin, IndependentPlugin):
+
+    short_desc = 'Nvidia(Mellanox) firmware tools output'
+
+    plugin_name = "mellanox_firmware"
+    profiles = ('hardware', 'system')
+
+    def setup(self):
+        # Get all devices which have the vendor Mellanox Technologies
+        devices = []
+        device_list = self.collect_cmd_output('lspci -D -d 15b3::0200')
+        # Will return a string of the following format:
+        # 0000:08:00.0 Ethernet controller: Mellanox Technologies MT2892
+        # Family
+        if device_list['status'] != 0:
+            # bail out if there no Mellanox PCI devices
+            return
+
+        for line in device_list["output"].splitlines():
+            # Should return 0000:08:00.0
+            # from the following string
+            # 0000:08:00.0 Ethernet controller: Mellanox Technologies MT2892
+            # Family
+            devices.append(line[0:8]+'00.0')
+
+        devices = set(devices)
+
+        # Collect info from certain commands, only if mft package is present
+        # Mft package is present if OFED is installed
+        mft_pred = SoSPredicate(self, packages=['mft'])
+        mft_mlxfw_pred = SoSPredicate(self,
+                                      packages=['mft'],
+                                      kmods=['mlx5_core', 'mlxfw'],
+                                      required={'kmods': 'all'})
+
+        # Collect info from certain commands, only if mstflint package is
+        # present.
+        # mstflint is part of the distro and can be installed.
+        # It will be overwritten once OFED is installed.
+        co = {'cmd': 'which flint', 'output': '/usr/bin/flint'}
+        mstflint_pred = SoSPredicate(self, packages=['mstflint'],
+                                     cmd_outputs=co,
+                                     required={'cmd_outputs': 'none'})
+        mstflint_mlxfw_pred = SoSPredicate(self,
+                                           packages=['mstflint'],
+                                           kmods=['mlx5_core', 'mlxfw'],
+                                           cmd_outputs=co,
+                                           required={'kmods': 'all',
+                                                     'cmd_outputs': 'none'})
+
+        for device in devices:
+            self.add_cmd_output("mlxdump -d %s pcie_uc --all" % device,
+                                pred=mft_pred)
+            self.add_cmd_output("mstconfig -d %s -e q" % device,
+                                pred=mft_pred)
+            self.add_cmd_output("flint -d %s dc" % device,
+                                pred=mft_pred)
+            self.add_cmd_output("flint -d %s q" % device,
+                                pred=mft_pred)
```

Comment:
Generally speaking it is fine for sos plugins to call `add_cmd_output()` without a predicate for commands that don't exist on a system - we handle those gracefully and don't pollute the plugin directory with empty output.

This is particularly why I say using the `packages` tuple for the plugin is likely a better option here.

Reply:
I guess we also would advise f-strings here?

Reply:
Makes sense. 
That means I could safely get rid of the predicates: `mft_mlxfw_pred` and `mstflint_mlxfw_pred` since they further add constraints to `mft_pred` and `mstflint_pred`?

I also switched to f-strings. Hopefully I did not miss any lines.